adam_core/coordinates/covariances.py

```diff
@@ -27,49 +29,141 @@
+    sigmas : `numpy.ndarray` (N, D)
+        Standard deviations for N coordinates in D dimensions.
 
     Returns
     -------
-    covariances : `~numpy.ma.masked_array` (N, D, D)
-        Covariance matrices with the squared 1-sigma values inserted along
-        each N diagonal.
+    covariances : `numpy.ndarray` (N, D, D)
+        Covariance matrices for N coordinates in D dimensions.
     """
-    if isinstance(sigmas, (np.ma.masked_array)):
-        sigmas_ = sigmas.filled()
-    else:
-        sigmas_ = sigmas
+    D = sigmas.shape[1]
+    identity = np.identity(D, dtype=sigmas.dtype)
+    covariances = np.einsum("kj,ji->kij", sigmas**2, identity, order="C")
+    return covariances
 
-    N, D = sigmas_.shape
-    if np.all(np.isnan(sigmas_)):
-        covariances = np.ma.zeros(
-            (N, D, D),
-            dtype=np.float64,
-        )
-        covariances.fill_value = COVARIANCE_FILL_VALUE
-        covariances.mask = np.ma.ones((N, D, D), dtype=bool)
 
-    else:
-        I = np.identity(D, dtype=sigmas_.dtype)  # NOQA: E741
-        covariances = np.einsum("kj,ji->kij", sigmas_**2, I)
-        covariances = np.ma.masked_array(
-            covariances,
-            dtype=np.float64,
-            fill_value=COVARIANCE_FILL_VALUE,
-        )
-        covariances.mask = np.ma.zeros((N, D, D), dtype=bool)
+class CoordinateCovariances(Table):
+    # TODO: Would be interesting if the dimensionality can be generalized
+    #      to D dimensions, so (N, D, D) instead of (N, 6, 6). We would be
+    #      able to use this class for the covariance matrices of different
+    #      measurments like projections (D = 4) and photometry (D = 1).
+    values = Field(pa.fixed_shape_tensor(pa.float64(), (6, 6)))
+
+    @property
+    def sigmas(self):
+        cov_diag = np.diagonal(self.to_matrix(), axis1=1, axis2=2)
+        sigmas = np.sqrt(cov_diag)
+        return sigmas
+
+    def to_matrix(self) -> np.ndarray:
+        """
+        Return the covariance matrices as a 3D array of shape (N, 6, 6).
+
+        Returns
+        -------
+        covariances : `numpy.ndarray` (N, 6, 6)
+            Covariance matrices for N coordinates in 6 dimensions.
+        """
+        return self.values.combine_chunks().to_numpy_ndarray()
+
+    @classmethod
+    def from_matrix(cls, covariances: np.ndarray) -> "CoordinateCovariances":
+        """
+        Create a Covariances object from a 3D array of covariance matrices.
+
+        Parameters
+        ----------
+        covariances : `numpy.ndarray` (N, 6, 6)
+            Covariance matrices for N coordinates in 6 dimensions.
+
+        Returns
+        -------
+        covariances : `Covariances`
+            Covariance matrices for N coordinates in 6 dimensions.
+        """
+        cov = pa.FixedShapeTensorArray.from_numpy_ndarray(covariances)
+        return cls.from_arrays([cov])
+
+    @classmethod
+    def from_sigmas(cls, sigmas: np.ndarray) -> "CoordinateCovariances":
```

Comment:
Sometimes you really just don't want to define all the elements in a covariance matrix and instead just create one from the 1-sigma uncertainties.